tsl/src/continuous_aggs/materialize.c

```diff
@@ -330,404 +413,368 @@ build_merge_update_clause(List *column_names)
-													  materialization_table,
-													  time_column_name,
-													  invalidation_range,
-													  chunk_condition->data);
+		char *query = create_materialization_plan_query(context, plan_type);
+		Oid types[] = { context->materialization_range.type, context->materialization_range.type };
+
+		materialization->plan = SPI_prepare(query, 2, types);
+		if (materialization->plan == NULL)
+			elog(ERROR, "%s: SPI_prepare failed: %s", __func__, query);
+
+		SPI_keepplan(materialization->plan);
 	}
+}
 
-	/* Get the max(time_dimension) of the materialized data */
-	if (rows_processed > 0)
+static void
+create_materialization_plans(MaterializationContext *context, MaterializationOperation type)
+{
+	switch (type)
 	{
-		spi_update_watermark(mat_ht,
-							 materialization_table,
-							 time_column_name,
-							 invalidation_range,
-							 chunk_condition->data);
+		case OPERATION_INSERT_DELETE:
+			create_materialization_plan(context, PLAN_TYPE_INSERT);
+			create_materialization_plan(context, PLAN_TYPE_DELETE);
+			break;
+		case OPERATION_MERGE_DELETE:
+			create_materialization_plan(context, PLAN_TYPE_INSERT);
+			create_materialization_plan(context, PLAN_TYPE_EXISTS);
+			create_materialization_plan(context, PLAN_TYPE_MERGE);
+			create_materialization_plan(context, PLAN_TYPE_MERGE_DELETE);
+			break;
+		default:
+			pg_unreachable();
+			break;
 	}
 }
 
-static bool
-spi_exists_materializations(SchemaAndName materialization_table, const NameData *time_column_name,
-							TimeRange materialization_range)
+static int
+execute_materialization_plan(MaterializationContext *context, MaterializationPlanType plan_type,
+							 uint64 *rows_processed)
 {
-	int res;
-	StringInfo command = makeStringInfo();
-	Oid types[] = { materialization_range.type, materialization_range.type };
-	Datum values[] = { materialization_range.start, materialization_range.end };
+	MaterializationPlan *materialization = get_materialization_plan(plan_type);
+	Datum values[] = { context->materialization_range.start, context->materialization_range.end };
 	char nulls[] = { false, false };
 
-	appendStringInfo(command,
-					 "SELECT 1 FROM %s.%s AS M "
-					 "WHERE M.%s >= $1 AND M.%s < $2 "
-					 "LIMIT 1;",
-					 quote_identifier(NameStr(*materialization_table.schema)),
-					 quote_identifier(NameStr(*materialization_table.name)),
-					 quote_identifier(NameStr(*time_column_name)),
-					 quote_identifier(NameStr(*time_column_name)));
-
-	elog(DEBUG2, "%s", command->data);
-	res = SPI_execute_with_args(command->data,
-								2,
-								types,
-								values,
-								nulls,
-								true /* read_only */,
-								0 /* count */);
+	int res = SPI_execute_plan(materialization->plan, values, nulls, materialization->read_only, 0);
 
-	if (res < 0)
-		elog(ERROR,
-			 "could not check the materialization table \"%s.%s\"",
-			 NameStr(*materialization_table.schema),
-			 NameStr(*materialization_table.name));
+	*rows_processed = SPI_processed;
```

Comment:
Nit: Seems like a redundant variable since you can use `SPI_processed` directly.